### PR TITLE
feat: multiline long array

### LINKED ISSUE #1217: [Proposal] Array Items on Separate Lines Fixer
I often find myself putting array items on separate lines. Specially when there are a lot of items in the arrays. Maybe we could add a fixer for that? Maybe it exists already? Example below:

``` php
// Before
protected $dates = ['deleted_at', 'ended_at', 'started_at'];

// After
protected $dates = [ 
    'deleted_at', 
    'ended_at', 
    'started_at'
];
```

What do you think?

### PR DESCRIPTION
**Pull request by bachinblack**
Fixes #1217 

---

This is my very first PR on an open source project so please be patient :smile: 
Also, it's possible that my PR is a duplicate of an already existing fixer, in which case I'll probably cry.

This PR is a possible solution to #4485 (issue I stumbled upon while looking for such a fixer).

It formats arrays according to the `max_length` argument (which is the maximum authorized length for a single line array).

## Why I think it's good.

I recently used [rector](https://github.com/rectorphp/rector) to upgrade my code and it turned my attributes into gigantic one-liners. Definitely not fixing them by hand.
Apart from that, I could use some logic on my decision-making regarding array formats. 

## Scope
1) The `max_length` argument is defined in characters and excludes whitespaces. Another possible option would be to count the number of items (not for me, though).

2) In the issue (#4485) it is also proposed to make an array multiline if it contains key/value pairs. To me it's not a great option as the type of an array hardly relates to its size but it's only fair to mention it.

3) It only works with arrays but it could also be useful for function arguments and chained method calls.
Possibly in the same fixer, given extra configuration (the code would be very close), possibly in separate fixers.

## Fixer priority (50)
This fixer leans on `ArrayIndentationFixer`, `WhitespaceAfterCommaInArrayFixer`, `NoWhitespaceBeforeCommaInArrayFixer`,`TrailingCommaInMultilineFixer` and `NoTrailingCommaInSinglelineFixer` to output nicely formatted code. It contains no logic at all regarding indentation or commas, is it okay ? Or should I also add this kind of logic to it ? Feels like duplication so I guess not.

Thanks for your time !

### CODE REVIEW

**Comment by bachinblack** [comment]
Fixed most of the problems with the CI but I can't seem to find how to test `getPriority()`. Can anyone help me with this ?

**Comment by Wirone** [comment]
@bachinblack hi, thank you for the contribution. From top of my head I can't tell right now if we already had such a fixer, but I'm on vacation now and can look deeper later (Thursday+). Anyway, I accepted workflow so you'll get CI feedback. In terms of testing priority you need integration test like [this](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/blob/master/tests/Fixtures/Integration/priority/align_multiline_comment%2Cphpdoc_trim_consecutive_blank_line_separation.test). 

PS. Please don't cry if it doesn't get merged, it's nice knowledge anyway (what you have learned during the process) 😊.

**Comment by bachinblack** [comment]
Hey ! Thanks for your help, I managed to add integration tests for my fixer.
I think it would be nice to add a section in the cookbook about priority tests. Maybe a subject for another PR ?
Anyway, it should be ready for a review and I'm open to input.

File: tests/Fixer/ArrayNotation/MultilineLongArrayFixerTest.php [hunk 1]
```diff
@@ -0,0 +1,437 @@
...
+    {
+        $this->fixer->configure($config);
+
+        $this->doTest($expected, $input);
+    }
+
+    public static function provideFixCases(): array
+    {
+        return [
+            // Old style array
```
Comment (Wirone) [inline tests/Fixer/ArrayNotation/MultilineLongArrayFixerTest.php:43]: Please use `yield 'old style array' => [ ... ];` for each case (requires `iterable` return type instead of an `array`). It's better from performance point of view (memory usage), but also cases' descriptions can be displayed when running with `--testdox` option. [See existing provider example](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/blob/12792fa271ad216530a7916fca91e397f7fe1ff4/tests/Fixer/Alias/ArrayPushFixerTest.php#L36-L39).

File: tests/Fixer/ArrayNotation/MultilineLongArrayFixerTest.php [hunk 2]
```diff
@@ -0,0 +1,437 @@
...
+                ['max_length' => 40],
+            ],
+            // don't change anonymous class implements list but change array inside
+            [
+                <<<'EXPECTED'
+            <?php
+            $x = [
+            1,
+            "2",
+            "c" => new class implements Foo ,Bar { const FOO = [
```
Comment (Wirone) [inline tests/Fixer/ArrayNotation/MultilineLongArrayFixerTest.php:406]:
```suggestion
            "c" => new class implements Foo, Bar { const FOO = [
```

Comma should be placed differently 😉, but this is only side comment - I am wondering if arrays like here should be refactored 🤔.

Comment (bachinblack) [reply tests/Fixer/ArrayNotation/MultilineLongArrayFixerTest.php:406]: Nice catch. As for your second comment, it's a good point. We could skip one-line stuff like classes, functions, arrow functions, conditions, ternary ?

Comment (Wirone) [reply tests/Fixer/ArrayNotation/MultilineLongArrayFixerTest.php]: It's always safer to start with smaller range and then add new cases rather than introduce something that later needs to be maintained for backward compatibility. I would like to know what others think about it: @kubawerlos, @keradus, @julienfalque?

Comment (bachinblack) [reply tests/Fixer/ArrayNotation/MultilineLongArrayFixerTest.php]: I'm not sure what you mean, should we format everything (like how it is now) or should we skip as many special cases as we can ?

Comment (Wirone) [reply tests/Fixer/ArrayNotation/MultilineLongArrayFixerTest.php]: I mean we should discuss it properly now and decide if test cases provided by you are OK because if we merge it, and people will start using, it will be harder to change behavior later 🙂. So if we decide that it's better to not touch some of these structures, then fixer has to be updated before merging.

File: doc/rules/array_notation/multiline_long_array.rst [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+Rule ``multiline_long_array``
+=============================
+
+A single-line array should be broken into multiple lines if it exceeds
+configured limit.
+
+Configuration
+-------------
+
+``max_length``
```
Comment (kubawerlos) [inline doc/rules/array_notation/multiline_long_array.rst:11]: Isn't that like common practice in the whole universe that the length of an array is the number of elements? Since `max_length` here is not can we have a good description of what is it?

File: tests/Fixer/ArrayNotation/MultilineLongArrayFixerTest.php [hunk 4]
```diff
@@ -0,0 +1,434 @@
...
+    {
+        $this->fixer->configure($config);
+
+        $this->doTest($expected, $input);
+    }
+
+    public static function provideFixCases(): iterable
+    {
+        yield 'Old style array' => [
+            <<<'EXPECTED'
```
Comment (kubawerlos) [inline tests/Fixer/ArrayNotation/MultilineLongArrayFixerTest.php:43]: Can we not have nowdocs in test cases? It's unnecessary noise - single quotes (and double in code) would make just fine.

File: tests/Fixer/ArrayNotation/MultilineLongArrayFixerTest.php [hunk 5]
```diff
@@ -0,0 +1,434 @@
...
+        yield 'Single-line array shorter than max_length' => [
+            <<<'EXPECTED'
+            <?php
+            $foo = ['foo','bar' => 'baz',];
+            EXPECTED,
+            null,
+            ['max_length' => 30],
+        ];
+
+        yield 'Single-line array with negative max_length' => [
```
Comment (kubawerlos) [inline tests/Fixer/ArrayNotation/MultilineLongArrayFixerTest.php:111]: There is no need for case testing some code is not changing if there is a test where the same code occurs as expected (like this one in the case `Multi line array with negative max_length`) - expected code is always checked that is no changing.

File: tests/Fixer/ArrayNotation/MultilineLongArrayFixerTest.php [hunk 6]
```diff
@@ -0,0 +1,434 @@
...
+            $y
+            ];
+            EXPECTED,
+            <<<'INPUT'
+            <?php
+            $x = [1, "2", "c" => new class implements Foo, Bar { const FOO = ["x","y"]; },$y ];
+            INPUT,
+            ['max_length' => 15],
+        ];
+    }
```
Comment (kubawerlos) [inline tests/Fixer/ArrayNotation/MultilineLongArrayFixerTest.php:433]: Can we have a test for input array like:
```php
$letters = ['a', /* @todo: add some other letters one day */ 'z'];
```
and `max_length` of `20`?

**Review by kubawerlos — DISMISSED** [review]

**Comment by Wirone** [comment]
FYI: there is issue #6225 and there was PR #6387 that addresses same concept.

File: tests/Fixer/ArrayNotation/MultilineLongArrayFixerTest.php (hunk 6 above)
Comment (bachinblack) [reply tests/Fixer/ArrayNotation/MultilineLongArrayFixerTest.php:433]: Good Idea.
I've added the test but I'm not sure about what we want it to return.

Here is the current output :
```php
<?php
$letters = [
"a",
/* @todo: add some other letters one day */ "z"
];
```

I think one of these would be more appropriate :
```php
<?php
$letters = [
"a", /* @todo: add some other letters one day */
"z"
];

$letters = [
"a",
/* @todo: add some other letters one day */
"z"
];
```
In any case, it could break users' habits when adding comments to arrays so another solution would be to skip comments entierly (right now they are included in max_length).
Any suggestions ?

Comment (bachinblack) [reply tests/Fixer/ArrayNotation/MultilineLongArrayFixerTest.php:433]: There are also comments in multi-line arrays, which are intended to be put in other places :
```php
<?php
$letters = [
"a", // This is the letter a
"z", /** This is not the letter a */
];

// becomes
<?php
$letters = [
"a",
// This is the letter a
"z",
/** This is not the letter a */
];

// Acceptable solutions could be 
<?php
$letters = [
// This is the letter a
"a",
/** This is not the letter a */
"z",
];

<?php
$letters = [
"a", // This is the letter a
"z", /** This is not the letter a */
];
```

Comment (bachinblack) [reply tests/Fixer/ArrayNotation/MultilineLongArrayFixerTest.php:354]: Here's how #6387 handles comments (I think it's equally unsatisfying) :
```php
<?php $a5 = [1,2,3,4,     /* 0 */5,/* 1 */ /* 2 */  6,/* 6+ */7,];

// becomes 
<?php $a5 = [
1,
2,
3,
4,     /* 0 */
5,/* 1 */ /* 2 */
6,/* 6+ */
7,
];
```

I believe there are 2 reasons someone would put comments inside an array :
- To explain values (see previous comment)
- - It mostly applies to multi-line arrays so I don't think we should change the shape of these arrays.
- To temporarily remove an element (see code above)
- - It would be possible to treat these as regular elements and apply formatting rules.

In any case, we would need to replace `//` with `/** */` for single-line arrays.

Since I don't see a reliable way of differenciating these two cases, I think the best solution would be to skip arrays with comments altogether, which is the change I just made to the PR.

File: src/Fixer/ArrayNotation/MultilineLongArrayFixer.php [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+                new CodeSample("<?php\n\$array = ['a very very long element','another very long element'];\n"),
+                new CodeSample("<?php\n\$array = ['a very very long element','another very long element'];\n", ['max_length' => 10]),
+            ]
+        );
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * Must run before ArrayIndentationFixer, NoTrailingCommaInSinglelineFixer, TrailingCommaInMultilineFixer, WhitespaceAfterCommaInArrayFixer.
```
Comment (bachinblack) [inline src/Fixer/ArrayNotation/MultilineLongArrayFixer.php:57]: This fixer should also come before `binary_operator_spaces`.

File: src/Fixer/ArrayNotation/MultilineLongArrayFixer.php [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+     */
+    public function getPriority(): int
+    {
+        return 50;
+    }
+
+    public function isCandidate(Tokens $tokens): bool
+    {
+        return $tokens->isAnyTokenKindsFound([T_ARRAY, CT::T_ARRAY_SQUARE_BRACE_OPEN]);
```
Comment (Wirone) [inline src/Fixer/ArrayNotation/MultilineLongArrayFixer.php:66]: I believe fixer should react also to `CT::T_DESTRUCTURING_SQUARE_BRACE_OPEN` and handle code like:

```php
// Before
[$foo, $bar, $baz] = $arr;

// After
[
    $foo,
    $bar,
    $baz
] = $arr;
```

Comment (bachinblack) [reply src/Fixer/ArrayNotation/MultilineLongArrayFixer.php]: Makes sense. I added the logic and I made a test for it, does it work for you ?

Comment (Wirone) [reply src/Fixer/ArrayNotation/MultilineLongArrayFixer.php]: Please add such case to `tests/Fixtures/Integration/priority/multiline_long_array,array_indentation.test` too, so we can be sure that it works as expected 🙂.

Comment (bachinblack) [reply src/Fixer/ArrayNotation/MultilineLongArrayFixer.php]: Will try to do it when I find some time this week

**Comment by Wirone** [comment]
@bachinblack in general it looks OK for me, but we had discussion about it with @kubawerlos and we think @SpacePossum's approach in #6387 was interesting, to have 2 separate conditions for triggering rule and strategy to choose (at least one, both). Your PR is focused on array length, but can be extended later with new options - are you interested in this? If yes, would you prefer to do it in this PR or later in second one? Of course we can discuss everything before so the outcome is as expected.

**Comment by bachinblack** [comment]
Hey @Wirone ! I wouldn't use this rule with an elements count myself because I don't think it's as reliable as a characters count but I guess it would be normal to give users the choice and it doesn't seem that hard to implement (though I believe there will be many more tests).

For the param names, I can get in line with @SpacePossum 's naming `elementsThreshold/charactersThreshold` and make them both nullable to let users pick what they want.

I think I'd be more comfortable merging it as it is now (after having renamed the current param to `charactersThreshold`) and starting a new one after if that's alright.

File: src/Fixer/ArrayNotation/MultilineLongArrayFixer.php (hunk 8 above)
Comment (bachinblack) [reply src/Fixer/ArrayNotation/MultilineLongArrayFixer.php]: I just added the test but I'm not sure what it was supposed to do.

I was expecting :
```php
// from
[$foo, $bar, $baz] = $arr;

// to
[
    $foo,
    $bar,
    $baz
] = $arr;
```
But I got :
```php
[
$foo,
$bar,
$baz
] = $arr;
```

After looking around, it seems like `ArrayIndentation` does not work with destructuring, right ?
So is this what was expected ? If not, is the test still relevant ?

**Comment by bachinblack** [comment]
As I proposed earlier, I rename `max_length` to `characters_threshold` so that it's clearer and it can easily be extended later with `elements_threshold`.

I understand fixers' default configuration can't just do nothing and there has to be a default config, so I chose to keep the default behaviour to `always multiline`.

Now, should we add `elements_threshold`, we'd need to keep this default behaviour.

File: src/Fixer/ArrayNotation/MultilineLongArrayFixer.php (hunk 8 above)
Comment (Wirone) [reply src/Fixer/ArrayNotation/MultilineLongArrayFixer.php]: It'll work when #7405 is merged, so we need to update integration test here when it's done 🙂.

Comment (bachinblack) [reply src/Fixer/ArrayNotation/MultilineLongArrayFixer.php]: Got you :+1: I just added the test

**Comment by bachinblack** [comment]
Hey ! I've caught up with master so it should be all good now, would anyone be so kind as to review my PR once more (hopfully it will be the last time haha) ?

**Comment by jklmnop** [comment]
This looks great seems like just what I'm looking for! Hoping it gets merged into release. 🤞

**Comment by mkroener** [comment]
I'm hoping this gets merged too as the current array formatting capabilities prevent me from fully integrating php-cs-fixer into my project.
If this is is abandoned and there is still interest in this I would be happy to work on this to get it merged

**Comment by bachinblack** [comment]
I'd be more than happy to merge it 😅 @kubawerlos, is there anything missing from this PR ? If not, would you be so kind as to accept it ? I can rebase it this week

File: tests/Fixer/ArrayNotation/MultilineLongArrayFixerTest.php (hunk 1 above)
Comment (jlecordier) [reply tests/Fixer/ArrayNotation/MultilineLongArrayFixerTest.php]: Same mistake there, I think we were both biased by the Cookbook which returns an array

**Comment by jlecordier** [comment]
See this [discussion](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/discussions/8040) for options idea

**Comment by jlecordier** [comment]
I will also add another idea here instead of creating yet another discussion :

Force function parameters to be on multiple lines when at least one of them has a default value, and is not the only parameter.

Before : 

```php
function my_function($param, $other = 'value')
{
   ...
}
```

After :

```php
function my_function(
    $param,
    $other = 'value'
) {
   ...
}
```

**Comment by jlecordier** [comment]
Just to be sure, is this issue supposed to resolve the following case ? :


Before : 

```php
$tab = [
    'a',
    'b', 'c'
];
```

After :

```php
$tab = [
    'a',
    'b',
    'c'
];
```

**Comment by bachinblack** [comment]
Hey @jlecordier!
You are correct, it would resolve the case you mentionned assuming you want small arrays to be multiline (based on the threshold you configure). 

Sadly, I'm not sure it will be accepted, it's been some time since I've created it and it hasn't moved since. 
I'm willing to rebase and even improve it if the maintainer(s) plan on accepting it.

As for your comment on function parameters, I think it's a great idea and have a similar implementation. It could probably be done in a single fixer.

**Comment by Wirone** [comment]
Sorry @bachinblack for not pushing this forward, but it's impossible to review and process everything, I need to prioritise things I do here as I have limited time. Parallel runner was most important for me, since it's merged I plan to slowly get back to the backlog. It would be great if you have rebased this, as we need to know if it satisfies our current QA suite anyway 🙂.

**Comment by jlecordier** [comment]
No worries @Wirone, take the time necessary !

I've been wondering why the team wants "a fixer to do it all" (a thing that works for array, and function, and arguments, and so on) ? Wouldn't it be more "agile" and simpler for everyone to do one thing at a time ? I know the code will probably be very similar for the different cases, but I think it would be easier to develop fixers independently, and then refactor later once each one is reviewed, so we can more easily identify duplicated code ?

**Comment by Wirone** [comment]
@jlecordier It's not like we want "a fixer to do it all". We like atomic, single-responsibility rules. They don't need to cover all the cases from day one, but similar logic should be put under one rule (like enforcing multiline in this case). There's no need to split it into `multiline_array`, `multiline_signature`, `multiline_function_call` etc. Just like we don't split FQCN shortening depending on where the symbol is determined, we do it in `fully_qualified_strict_types` (the name is invalid at this point, but we have it in backlog, so it will be improved in the future). Refactoring rules is not that easy and it does not make sense to introduce rule just to deprecate it shortly in favour of other rule - it's a waste of time for maintainers/contributors, but also for users who need to clean up their configs.

**Comment by jlecordier** [comment]
Mhh, I guess that's subjective then, thanks for answering anyway !

**Comment by bachinblack** [comment]
Hey @Wirone I just rebased and apparently I get an error in the CI. I've checked the logs and tried to reproduce in local, I can't find the cause of the failure. Do you see what's going on ? :thinking:

**Comment by bachinblack** [comment]
Hey @Wirone ! I rebased but I can't find the error I'm getting with the CI that's failing :( Could you help me figure out what's going on ?

**Comment by Wirone** [comment]
@bachinblack this is buggy test related to parallel runner, we need to improve it as it sometimes fails and need to be re-run. Please ignore it for now. More interesting is why "Static Code Analysis" workflow did not start / report status 🤔.

In general: sorry for the silence from my side, for long time I was focused on `#7777` and barely work on other PRs that demand high focus. I am slowly trying to get back to other stuff, but I have a lot of notifications to process and I need some time for it.

**Comment by nlighteneddesign** [comment]
Are there any tips on how to test this?
When I try to run this locally I get an error:

```
Class PhpCsFixer\Fixer\ArrayNotation\MultilineLongArrayFixer contains 2 abstract methods and must therefore be declared abstract or implement the remaining methods (PhpCsFixer\Fixer\ConfigurableFixerInterface::configure, PhpCsFixer\Fixer\ConfigurableFixerInterface::getConfigurationDefinition) in /var/www/html/vendor/friendsofphp/php-cs-fixer/src/Fixer/ArrayNotation/MultilineLongArrayFixer.php on line 32
```

I stubbed out that class to run it and got the following error:

```
Fatal error: Uncaught Error: Call to undefined method PhpCsFixer\AbstractFixer::configure() in /var/www/html/vendor/friendsofphp/php-cs-fixer/src/Fixer/ArrayNotation/MultilineLongArrayFixer.php:38
```

I then removed the call to `parent::configure($configuration);`
After this it ran, but didn't actually change the arrays.

Any advice would be appreciated.

File: src/Fixer/ArrayNotation/MultilineLongArrayFixer.php [hunk 9]
```diff
@@ -0,0 +1,240 @@
...
+use PhpCsFixer\FixerDefinition\FixerDefinition;
+use PhpCsFixer\FixerDefinition\FixerDefinitionInterface;
+use PhpCsFixer\Tokenizer\CT;
+use PhpCsFixer\Tokenizer\Token;
+use PhpCsFixer\Tokenizer\Tokens;
+
+/**
+ * @author Philippe Bouttereux <philippe.bouttereux@gmail.com>
+ */
+final class MultilineLongArrayFixer extends AbstractFixer implements ConfigurableFixerInterface
```
[the remark this answers is not in the document]
Comment (bachinblack) [reply src/Fixer/ArrayNotation/MultilineLongArrayFixer.php:32]: How about `arrayLinesFormat ` ?

File: src/Fixer/ArrayNotation/MultilineLongArrayFixer.php [hunk 10]
```diff
@@ -0,0 +1,240 @@
...
+    {
+        parent::configure($configuration);
+
+        $this->charactersThreshold = $this->configuration['characters_threshold'] ?? 0;
+    }
+
+    public function getDefinition(): FixerDefinitionInterface
+    {
+        return new FixerDefinition(
+            'A single-line array should be broken into multiple lines if it exceeds configured limit. Arrays that contain comments should be left unchanged.',
```
[the remark this answers is not in the document]
Comment (bachinblack) [reply src/Fixer/ArrayNotation/MultilineLongArrayFixer.php:46]: Would this fit better : `Arrays must be either singleline or multilines based on a given length threshold. Arrays containing comments shall be left unchanged.`

**Comment by nlighteneddesign** [comment]
Still an issue

File: src/Fixer/ArrayNotation/MultilineLongArrayFixer.php [hunk 11]
```diff
@@ -0,0 +1,240 @@
...
+            T_ARRAY,
+            CT::T_ARRAY_SQUARE_BRACE_OPEN,
+            CT::T_DESTRUCTURING_SQUARE_BRACE_OPEN,
+        ]);
+    }
+
+    protected function createConfigurationDefinition(): FixerConfigurationResolverInterface
+    {
+        return new FixerConfigurationResolver([
+            (new FixerOptionBuilder('characters_threshold', 'Maximum length in characters (excluding whitespaces) for single-line arrays. 0 : always multi-line, -1 : always single-line.'))
```
[the remark this answers is not in the document]
Comment (bachinblack) [reply src/Fixer/ArrayNotation/MultilineLongArrayFixer.php:76]: I did it in order to keep the `null` value free in case we want to add `items_threshold` (and therefore disable `characters_threshold`). Thinking about it, I guess we could define arguments another way : a param `threshold: ?int` and a param `unit: string(characters|items)`
By doing so, we make it impossible to stack both logics (by items and by characters) but I don't believe it's that important, but we make it so that every param is always necessary : `null` can be used to mean `infinity`.
I couldn't find any enums in the codebase, I'm assuming const strings are the way to go for defining such parameters ?

**Comment by bachinblack** [comment]
Hey ! Finally got around to rebasing and fixing this PR.
I changed the name to `ArraySinglelineMultiline` and improved the config. I also addressed all the remaining qa errors.
Let me know what you think.

Quick thought on the name, I also thought of `ArrayLineBreaks`, I could change it if you prefer.

File: src/Fixer/ArrayNotation/ArraySinglelineMultilineFixer.php [hunk 12]
```diff
@@ -0,0 +1,229 @@
...
+                new CodeSample("<?php\n\$array = ['a very very long element','another very long element'];\n"),
+                new CodeSample("<?php\n\$array = ['a very very long element','another very long element'];\n", ['max_length' => 10]),
+            ]
+        );
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * Must run before ArrayIndentationFixer, NoTrailingCommaInSinglelineFixer, TrailingCommaInMultilineFixer, WhitespaceAfterCommaInArrayFixer.
```
Comment (jlecordier) [reply src/Fixer/ArrayNotation/ArraySinglelineMultilineFixer.php:63]: Why not after ? The length could be below the threshold before all the other fixers, and above the threshold after the other fixers

File: doc/rules/array_notation/array_singleline_multiline.rst [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+You can configure this rule using the following option: ``threshold``.
+
+Configuration
+-------------
+
+``threshold``
+~~~~~~~~~~~~~
+
+Maximum length in characters (excluding whitespaces) for single-line arrays. 0 :
+always multiline, -1 : makes empty arrays multiline, null : always singleline.
```
Comment (jlecordier) [inline doc/rules/array_notation/array_singleline_multiline.rst:23]: There are 4 cases : 0, -1, null and n, but you only provide 2 examples
